operator/pkg/controlplane/control_plane_test.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"istio.io/istio/operator/pkg/util"
+
+	"istio.io/api/operator/v1alpha1"
+	"istio.io/istio/operator/pkg/component"
+	"istio.io/istio/operator/pkg/name"
+	"istio.io/istio/operator/pkg/translate"
+)
+
+func TestOrderedKeys(t *testing.T) {
+	tests := []struct {
+		desc string
+		in   map[string]*v1alpha1.ExternalComponentSpec
+		want []string
+	}{
+		{
+			desc: "not-ordered",
+			in: map[string]*v1alpha1.ExternalComponentSpec{
+				"graphql":   nil,
+				"Abacus":    nil,
+				"Astrology": nil,
+				"gRPC":      nil,
+				"blackjack": nil,
+			},
+			want: []string{
+				"Abacus",
+				"Astrology",
+				"blackjack",
+				"gRPC",
+				"graphql",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			if got := orderedKeys(tt.in); !(reflect.DeepEqual(got, tt.want)) {
+				t.Errorf("%s: got %+v want %+v", tt.desc, got, tt.want)
+			}
+		})
+	}
+}
+
+func TestNewIstioOperator(t *testing.T) {
+	coreComponentOptions := &component.Options{
+		InstallSpec: &v1alpha1.IstioOperatorSpec{},
+		Translator:  &translate.Translator{},
+	}
+	tests := []struct {
+		desc              string
+		inInstallSpec     *v1alpha1.IstioOperatorSpec
+		inTranslator      *translate.Translator
+		wantIstioOperator *IstioOperator
+		wantErr           error
+	}{
+		{
+			desc:          "core-components",
+			inInstallSpec: &v1alpha1.IstioOperatorSpec{},
+			inTranslator: &translate.Translator{
+				ComponentMaps: map[name.ComponentName]*translate.ComponentMaps{
+					"Pilot": {
+						ResourceName: "test-resource",
+					},
+				},
+			},
+			wantErr: nil,
+			wantIstioOperator: &IstioOperator{
+				components: []component.IstioComponent{
+					&component.CRDComponent{
+						CommonComponentFields: &component.CommonComponentFields{
+							Options:       coreComponentOptions,
+							ComponentName: name.IstioBaseComponentName,
+						},
+					},
+					&component.PilotComponent{
+						CommonComponentFields: &component.CommonComponentFields{
+							Options: &component.Options{
+								InstallSpec: &v1alpha1.IstioOperatorSpec{},
+								Translator:  &translate.Translator{},
+							},
```

Comment:
equal to 
```
Options:       coreComponentOptions,
```
?

Reply:
In addition, IMO we may need some more complicated cases, constructing IstioOperator object with more complicated Translator, especially gateways and addonComponents, as they are slice and map type.

Reply:
@morvencao that is a good idea, but I think I will do that in a separate PR. Is everything else OK? I'm still learning how all these components to operator work.

Reply:
sure, thanks!